# Numpy


In [16]:
import numpy as np

## Creating Numpy array from Python list


In [17]:
# 1 dimensional array
array = np.array([1, 2, 3, 4, 5])
print(array)

[1 2 3 4 5]


In [18]:
# fmt:off
# 2 dimensional array
array2D = np.array(
  [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
  ])
print(array2D)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


### Python list vs Numpy array

- Numpy array computations are much faster than python lists
- List multiplication will repeat elements
- Numpy array multiplication will perform element-wise multiplication


In [19]:
list = [1, 2, 3]
print("List Multiplication:", list * 2)

List Multiplication: [1, 2, 3, 1, 2, 3]


In [20]:
array = np.array([1, 2, 3])
print("Numpy array Scalar multiplication:", array * 2)
print("Numpy array vector multiplication:", array * [1, 2, 3])

Numpy array Scalar multiplication: [2 4 6]
Numpy array vector multiplication: [1 4 9]


### Array with range of values


In [21]:
array_range1 = np.arange(10)  # from 0 to 10 (end-1)
array_range2 = np.arange(5, 10)  # from 5 to 9 <start, stop>
array_range3 = np.arange(0, 10, 2)  # from 0 to 9 with step of 2 <start, stop, step>

print("Array ranging 0 - 9:", array_range1)
print("Array ranging 5 - 9:", array_range2)
print("Array ranging 0 - 9 with step of 2:", array_range3)

Array ranging 0 - 9: [0 1 2 3 4 5 6 7 8 9]
Array ranging 5 - 9: [5 6 7 8 9]
Array ranging 0 - 9 with step of 2: [0 2 4 6 8]


### Array with equally spaced value for a given range


In [22]:
array_spaced = np.linspace(0, 10, 5)
print("Equally spaced value array ranging 0 - 10:", array_spaced)

Equally spaced value array ranging 0 - 10: [ 0.   2.5  5.   7.5 10. ]


## Ways to Creating Matrix


### Zeros Matrix

$$
\text{zeros} =
\begin{bmatrix}
0 & 0 & 0 \\
0 & 0 & 0 \\
0 & 0 & 0
\end{bmatrix}
$$


In [23]:
zeros = np.zeros((3, 3))
print("Zero Matrix:\n", zeros)

Zero Matrix:
 [[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


### Ones Matrix

$$
\text{ones} =
\begin{bmatrix}
1 & 1 & 1 \\
1 & 1 & 1 \\
1 & 1 & 1
\end{bmatrix}
$$


In [24]:
ones = np.ones((3, 3))
print("Ones Matrix:\n", ones)

Ones Matrix:
 [[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


### Identity Matrix

$$
\text{I}_3 =
\begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1
\end{bmatrix}
$$


In [49]:
identity = np.identity(3)
print("Identity Matrix:\n", identity)

Identity Matrix:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


### Constant Matrix/ Uniform Matrix

$$
\text{constants} =
\begin{bmatrix}
2 & 2 & 2 \\
2 & 2 & 2 \\
2 & 2 & 2
\end{bmatrix}
$$


In [26]:
constants = np.full((3, 3), 2)
print("Constant Matrix/ Uniform Matrix:\n", constants)

Constant Matrix/ Uniform Matrix:
 [[2 2 2]
 [2 2 2]
 [2 2 2]]


### Random Matrix

$$
random =
\begin{bmatrix}
a_{11} & a_{12} & a_{13} \\
a_{21} & a_{22} & a_{23} \\
a_{31} & a_{32} & a_{33}
\end{bmatrix}
\quad \text{where} \quad a_{ij} \in [0,1)
$$


In [27]:
random = np.random.random((3, 3))
print("Random Matrix:\n", random)

Random Matrix:
 [[0.67276755 0.84736895 0.55394181]
 [0.72326304 0.33109153 0.43372646]
 [0.2756536  0.22806488 0.27767116]]


### Empty Matrix


In [31]:
empty = np.empty((3, 3))
print("Empty Matrix:\n", empty)

Empty Matrix:
 [[4.9e-324 9.9e-324 1.5e-323]
 [2.0e-323 2.5e-323 3.0e-323]
 [3.5e-323 4.0e-323 4.4e-323]]


## Vector, Matrix, and Tensor

$$
\begin{aligned}
\text{Vector:} \quad \mathbf{V} &=
\begin{bmatrix}
x_1 & x_2 & x_3
\end{bmatrix}
\\[2ex]

\text{Matrix:} \quad \mathbf{M} &=
\begin{bmatrix}
a_{11} & a_{12} & a_{13} \\
a_{21} & a_{22} & a_{23} \\
a_{31} & a_{32} & a_{33}
\end{bmatrix}
\\[5ex]

\text{Tensor:} \quad \mathcal{T} &=
\left[
\begin{array}{ccc}
a_{111} & a_{112} & a_{113} \\
a_{121} & a_{122} & a_{123} \\
a_{131} & a_{132} & a_{133} \\
\end{array}
\right]
\quad
\left[
\begin{array}{ccc}
a_{211} & a_{212} & a_{213} \\
a_{221} & a_{222} & a_{223} \\
a_{231} & a_{232} & a_{233} \\
\end{array}
\right]
\quad
\left[
\begin{array}{ccc}
a_{311} & a_{312} & a_{313} \\
a_{321} & a_{322} & a_{323} \\
a_{331} & a_{332} & a_{333} \\
\end{array}
\right]
\end{aligned}
$$


In [28]:
# fmt: off
vector = np.array([10, 20, 30])
print("Vector:\n", vector)

matrix = np.array(
    [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
    ]
)
print("Matrix:\n", matrix)

tensor = np.array(
    [
        [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9],
        ],
        [
            [10, 20, 30],
            [40, 50, 60],
            [70, 80, 90],
        ],
        [
            [100, 200, 300],
            [400, 500, 600],
            [700, 800, 900],
        ],
    ]
)
print("Tensor:\n", tensor)

Vector:
 [10 20 30]
Matrix:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Tensor:
 [[[  1   2   3]
  [  4   5   6]
  [  7   8   9]]

 [[ 10  20  30]
  [ 40  50  60]
  [ 70  80  90]]

 [[100 200 300]
  [400 500 600]
  [700 800 900]]]


## Numpy Array properties:


In [29]:
array = np.array(
    [  # fmt:skip
        [1, 2, 3],
        [4, 5, 6],
    ]
)
print("Shape:", array.shape)
print("Size:", array.size)
print("Dimension:", array.ndim)
print("Data Type:", array.dtype)
print("Item Size:", array.itemsize, "bytes")

Shape: (2, 3)
Size: 6
Dimension: 2
Data Type: int64
Item Size: 8 bytes


### Changing datatype:

- Returns new array with new Data Type
- Does not affects original array


In [30]:
age = np.array([7, 18, 36, 23, 75, 60, 53, 24, 68, 28, 34, 32])
print("Data Type of Age:", age.dtype)
age_int8 = age.astype(np.uint8)
print("Data Type of Age (int8):", age_int8.dtype)
print("Data Type of Age after Changing:", age.dtype)

Data Type of Age: int64
Data Type of Age (int8): uint8
Data Type of Age after Changing: int64
